### DISCUSSION TITLE
Lets talk about the pose-widget for a moment.

### DISCUSSION THREAD

**Discussion by nuxil** [opentrack/opentrack#1418] [Show and tell]
Since i discovered this program "not so long ago", i have become found of it.
I used to have a TrackIR5 but its broken, i found this program to be equal as good. if not better :)
Now this is my goto program when i want to play DCS .

But one thing that has bothered me since i started using it, and thats the pose-widget.
I dont like the octopussy, dont get me wrong. its a nice image. 
But its not symertical.  (maybe im nitpicking),  ​there is no good feel of depth in the widget
The mirror check box is unnecessary in my opinion, since we can already invert the axis in the options.

The last couple of days i been trying to wrap my head around the code in this project.
So i can do some changes i like for my personal build.

I have been changing the 2d widget into a 3d widget.
It is still work in progress. but at this point. its mostly tweeking the 3dmodels and its textures.

But i would like opinions from others. 
Is this something i should submit a git pull request for when its done? 

What are your thought?

![image](https://i.imgur.com/3LU9iBB.png)

**Comment by sthalik** [comment]
Do I understand correctly that you managed to implement `QGLWidget` with a passthrough shader? Big if true.

**Comment by GO63-samara** [comment]
I think this way about improving pose-widget:

1. I still thought that the 2D widget is used in order not to increase CPU usage on weak computers. 3D looks good and might be fine as long as it doesn't increase CPU usage.
2. The octopus has been the project's mascot for a long time and should definitely be kept by default. It is possible to make a selection of other 3D models in options.
3. In the proposed widget, I do not like the checkered background. It reminds me of a zoo or a prison. Quite enough basic axes.
4. I think a mirror check box is needed. It does not affect the output coordinates. With a mirror, it is more convenient to correlate the movements of your real head with the movements of an octopus. It might make sense to remove the mirror check box from pose-widget to the options window. There it is more convenient to save its value in ini.
5. I still think it would be very useful for trackers without video to output another input pose-widget instead of the "NO VIDEO" alert.

Do you agree with me? What are your thought?

**Comment by nuxil** [comment]
_> Do I understand correctly that you managed to implement `QGLWidget` with a passthrough shader? Big if true._
@sthalik Well. its just Qt3D that i have used. not really special about it.

@GO63-samara

1. ofcource going from 2D to 3D will increase the Cpu load a bit more. how much also depends on the model used.
But having a increcement on the cpu usage at 0.1% with 2D to 0.15-0.3% with 3D (in my case). That is not a big issue.
Memory will also increase a bit. around 20MB, but that also depends a bit on the model used.
How many pollys it got. texture size etc.

More so, today. cpu's have multiple cores. My cpu isnt the newest, a old Ryzen with 12 cores. 24 logical cores.
So yea, running opentrack on one of the cores wouldnt cause any performance issue.

And lets not forget about games. games to rarly use more than 2 physical  cores. they may use many threads but not many cores.
Due to issues with synchronization between them and all sort of other issues.

2. I recond that the octopusy was the logo | mascot.  
I guess it would be possible to create a option to select between types of pose widgets. 2D or 3D

3. its not a checker board texture.. its a wireframe texture, a  small difference :)
As i stated in my 1st post.  This is still work in progressm, i too dont like it much and i will change it to something else.

4. If not to be removed, i agreed on moving it to the option menu. 
but then again. (translated expression) that is just lard on butter.
meaning that since the options already exist to invert axis. 
its not really needed and might just clutter things more up and cause confusion.

5. I dont mind the No Video.  not sure what you meant about another input pose-widget
Are you taking about replacing both the pose widget and the video widget with something else if people use. example UDP?

**Comment by sthalik** [comment]
@nuxil:

The pose shouldn't refresh when the window is minimized or occluded. Either draw it to a framebuffer during update and blit that onto the widget's surface during redraws, or just avoid the redraws altogether while the window's not visible.

**Comment by GO63-samara** [comment]
1.  I agree, CPU usage will increase imperceptibly. You can try to implement a 3D widget.

> 2.  I guess it would be possible to create a option to select between types of pose widgets. 2D or 3D

No, it's redundant. If you are already switching to 3D, then 2D does not need to be pulled with you. If the 3D model is in a separate file plus a texture file, then the user can replace the 3D model with his own. That might be easier to implement.

> 3. As i stated in my 1st post. This is still work in progressm, i too dont like it much and i will change it to something else.

Now the screenshots look pretty gloomy. It's like diffuse lighting. I would add a couple of directional lights. One light from the camera, another light from above. This would add volume and realism.

> 4. If not to be removed, i agreed on moving it to the option menu.
> but then again. (translated expression) that is just lard on butter.
> meaning that since the options already exist to invert axis.
> its not really needed and might just clutter things more up and cause confusion.

No, it's not the same thing. Inverting the axes inverts the output coordinates. And the mirror checkbox only affects the rendering of the pose-widget. Imagine how comfortable it is to comb while looking at the webcam image on the monitor or looking in the mirror?

> 5.  Are you taking about replacing both the pose widget and the video widget with something else if people use. example UDP?

I mean add another exactly the same pose-widget in place of "NO VIDEO" for non-video trackers like UDP or Hatire Arduino. The left pose-widget is the data visualization of the "Raw tracker data" panel, and the right pose-widget is already there, this is the data visualization of the "Game data" panel.

> or just avoid the redraws altogether while the window's not visible.

I like it. This is the most correct solution with CPU usage.

**Comment by GO63-samara** [comment]
I recently made a 3D model of a "gingerbread octopus". You can try if it works for you:

https://sites.google.com/site/diyheadtracking/устройства-обзора-для-авиасимуляторов/opentrack/dummytrack-6dof-симулятор-трекера-и-6dof-индикатор-позы
https://drive.google.com/file/d/1JTiiRFwsXyEBGyas9ruGwZGzYG3A0TnA/view

You need files from the archive Octopus.3DS and Octopus.jpg

**Comment by sthalik** [comment]
> @nuxil wrote:
>> I guess it would be possible to create a option to select between types of pose widgets. 2D or 3D

@GO63-samara wrote:
> If you are already switching to 3D, then 2D does not need to be pulled with you. 

Agreed. Either it's good enough to be merged and supersede the old version or not.

Reducing the CPU and GPU overhead while the window is minimized, is a critical concern. Slight CPU usage increase while the window is focused, not so much. Ideally, `paintEvent` or equivalent won't even be called when hidden. If that leaves artifacts, so be it.

> I mean add another exactly the same pose-widget in place of "NO VIDEO" for non-video trackers like UDP or Hatire Arduino. The left pose-widget is the data visualization of the "Raw tracker data" panel, and the right pose-widget is already there, this is the data visualization of the "Game data" panel.

This could be done even now as the 2D widget is reentrant. Depends on what the users of camera-less trackers think.

> Now the screenshots look pretty gloomy. It's like diffuse lighting. I would add a couple of directional lights. One light from the camera, another light from above. This would add volume and realism.

As long as the OpenGL 3.3+ API (or GLES, depends on what Qt uses) and not the fixed pipeline is used, extending the widget for any particular aesthetics shouldn't be a problem.

**Comment by nuxil** [comment]
@GO63-samara 

**_"Now the screenshots look pretty gloomy. It's like diffuse lighting. I would add a couple of directional lights. One light from the camera, another light from above. This would add volume and realism."_**

Screenshot is "gloomy". 
It's just test models i have.. and its bascially  a10min paint job done on the textures.
No normal maps, bump maps etc. just a plain color texture.
The textures need heavly improvement.  I need make normal maps etc. when i get the time.

Big No on the light sources. if you are conserned about gpu usage. defently not.    
Light sources will increase the load on the gpu. espesialy if shadows involved aswell. 

The textures themself will be the lightsource.  They will have ambient lighting.
Do we really need a fancy scene with lighting, shadows etc ? 

ofcource if you all want lightsources in the scene, i can make that.  but since performance has been a theme here,
I dont think its a good idea.


**_"No, it's not the same thing. Inverting the axes inverts the output coordinates. And the mirror checkbox only affects the rendering of the pose-widget"_**
gotcha.  now it make sense :p


**_"If the 3D model is in a separate file plus a texture file, then the user can replace the 3D model with his own. That might be easier to implement."_**

Yes. that has been one of the goals.
We could have a "opentrack/3dmodel" directory  and with fixed names. ie. head.obj and background.obj
idk. come up with better names if you like.

I have download your model. but i will need to convert it to obj format, since  this is what i am working with.


**_"I mean add another exactly the same pose-widget in place of "NO VIDEO" for non-video trackers like UDP or Hatire Arduino.  ...."_**
I'll leave that work for someone else.  at least for the time being.


@sthalik 
**_"Agreed. Either it's good enough to be merged and supersede the old version or not."_**
I'll try get as much done as possibel until the end of the weekend,  then you can check the code out.
But no promisises, i have a busy weekend.

**_"Reducing the CPU and GPU overhead while the window is minimized, is a critical concern."_**
Its only rendering on demans. Meaning, its only updating the scene when changes are done. 
So the gpu usage should be minimum. When the program is minimized. the gpu load is 0%.


Disable rendering if the window it getting overlapped is a more complex task.
It would need it own code for windows, linux and mac.
On Windows we would enumerate all windows that is above our program. "windows" are ordered by Z order on Windows OS.
Then compare the RECTS. 

But linux and mac would need its own implementation. i have no idea how you do it on thouse OS'es
So if you want to minimize the cpu and gpu load. one should always minimize the program

**_"As long as the OpenGL 3.3+ API (or GLES, depends on what Qt uses) and not the fixed pipeline is used, extending the widget for any particular aesthetics shouldn't be a problem."_**

Indeed. :)

**Comment by sthalik** [comment]
>> "If the 3D model is in a separate file plus a texture file, then the user can replace the 3D model with his own. That might be easier to implement."
> Yes. that has been one of the goals.
> We could have a "opentrack/3dmodel" directory and with fixed names. ie. head.obj and background.obj

I think embedding these in the executable is the better choice. The mesh as a list of vertices (float or 16-bit fixed-point) can even be embedded as a `.c` file to be more compact than `.obj`. There's precedent with `opentrack.png`.

> Its only rendering on demans. Meaning, its only updating the scene when changes are done.
> So the gpu usage should be minimum. When the program is minimized. the gpu load is 0%.

Great. That completely solves the performance question.

With regard to adding fancy effects -- going beyond ambient + diffuse without putting in serious skill and resources, will make things look worse. For instance, see how specular reflections look in free games derived from the DOOM 3 renderer. Human models look like a plate of gelatin standing in the rain with a flashlight pointing at it. So restricting the model to purely UV mapping + constant ambient light is a good thing. Less is more.

> Disable rendering if the window it getting overlapped is a more complex task.
> It would need it own code for windows, linux and mac.

It's working for Windows, see `compat/check-visible`. There's some Linux equivalent, see this Stack Overflow question <<https://archive.is/iZMoC>>. OSX doesn't even have fully working keyboard shortcuts support which are far more important. Let's leave it until someone steps up to finish the port.

OSX has awful 3D support anyway. It's worse than Mesa, perhaps even worse than the old fglrx. So much for gaming on Apple hardware.

**Comment by nuxil** [comment]
**_"I think embedding these in the executable is the better choice. The mesh as a list of vertices (float or 16-bit fixed-point) can even be embedded as a .c file to be more compact than .obj. There's precedent with opentrack.png."_**

Thats is actually not a bad idea.  
Should GO63-samara's model be the offical model to use then ? 

**_"With regard to adding fancy effects -- going beyond ambient + diffuse without putting in serious skill and resources, will make things look worse..."_**
This i kind of agree with. no need to overcomplicate things.

Anywho. we can all agree that the "Zoo" cage as GO63-samara calls it, is dead ugly. 
I did change the backround model a bit.

Here is a short video,  I used Recordit, (crap program), frames are not smooth in video but are in real life. sorry if it gives wrong impression.

https://recordit.co/J5aud3Y1lc

What do you think? 
If you have your own ideas about how it should look like. 
Please share them. 

![image](https://i.imgur.com/ptGwv5n.png)

**Comment by sthalik** [comment]
> Should GO63-samara's model be the offical model to use then ?

It could use more user-visible depth, just a little to establish the existence of the third dimension. I could be wrong though, have to try it one way or another. But the human head from the centermost image of your OP is good too.

> Here is a short video, I used Recordit, (crap program), frames are not smooth in video but are in real life. sorry if it gives wrong impression.

The distortion looks bad. The lines look as if they were ellipses around the screen's center. They don't look like axis-aligned lines or separate faces of a cube.

The background looks better than the padded room in the previous version. The color scheme could be more subdued, but these are merely compile-time constants so no need to argue about them too much now.

**Comment by GO63-samara** [comment]
> I have download your model. but i will need to convert it to obj format, since this is what i am working with.

I also prefer obj format for development. For publishing, I chose 3DS format because the file size is smaller and no additional mtl file is needed.


> "I mean add another exactly the same pose-widget in place of "NO VIDEO" for non-video trackers like UDP or Hatire Arduino. ...."
> I'll leave that work for someone else. at least for the time being.

Of course, this is more of a suggestion for @sthalik.


> Should GO63-samara's model be the offical model to use then ?

Of course it doesn't have to be official! I just showed how you can make a lowpoly 3D model look like an already existing 2D octopus in a widget. I'm sure you can make the model even better. But in my opinion it should still be the default octopus model.


> Here is a short video, I used Recordit, (crap program), frames are not smooth in video but are in real life. sorry if it gives wrong impression.

I use FastStone Capture to capture video. This works without delay and immediately compresses to wmv.
It's not clear in your video why the background stays in place when moving, but rotates with the model. Is this checkered sphere necessary? In my opinion, yellow axles are enough. These axes can pierce the 3D model, showing that the model is exactly in the center.


> I think embedding these in the executable is the better choice. The mesh as a list of vertices (float or 16-bit fixed-point) can even be embedded as a .c file to be more compact than .obj. There's precedent with opentrack.png.

I would leave separate file plus a texture file. Even now we don't all vote for the octopus. For the user, the ability to replace the model will be a nice bonus too.

**Comment by 5YNTH3T1K** [comment]
TLDR :  Anything come of this ?  Asking for a friend...

**Comment by 5YNTH3T1K** [comment]
#1928

**Reply by nuxil** [reply]
I am sorry.
was busy back then with work and real life stuff.
i forgot about this as time passed by.

not sure if i will pick up on it again as i had some issues compiling the neural plugin using Msys/Mingw toolchain. "it annoyed me a lot"
besides i was never happy with the artistic part of the 3D. 
but who knows what the future will bring, dont have high hopes tho.
